cli/azd/cmd/login.go

```diff
@@ -57,50 +56,58 @@ func loginCmdDesign(global *internal.GlobalCommandOptions) (*cobra.Command, *log
 }
 
 type loginAction struct {
-	formatter output.Formatter
-	writer    io.Writer
-	console   input.Console
-	azCli     azcli.AzCli
-	flags     loginFlags
+	formatter   output.Formatter
+	writer      io.Writer
+	console     input.Console
+	authManager *auth.Manager
+	flags       loginFlags
 }
 
 func newLoginAction(
 	formatter output.Formatter,
 	writer io.Writer,
-	azcli azcli.AzCli,
+	authManager *auth.Manager,
 	flags loginFlags,
 	console input.Console,
 ) *loginAction {
 	return &loginAction{
-		formatter: formatter,
-		writer:    writer,
-		console:   console,
-		azCli:     azcli,
-		flags:     flags,
+		formatter:   formatter,
+		writer:      writer,
+		console:     console,
+		authManager: authManager,
+		flags:       flags,
 	}
 }
 
-func (la *loginAction) Run(ctx context.Context) error {
-	if err := tools.EnsureInstalled(ctx, la.azCli); err != nil {
-		return err
-	}
+const (
+	// CodespacesEnvVarName is the name of the env variable set when you're in a Github codespace. It's
+	// just set to 'true'.
+	CodespacesEnvVarName = "CODESPACES"
+
+	// RemoteContainersEnvVarName is the name of the env variable set when you're in a remote container. It's
+	// just set to 'true'.
+	RemoteContainersEnvVarName = "REMOTE_CONTAINERS"
```

Comment:
Wondering if there would be a way to also detect `local container`, `remote-ssh` and `remote tunnels`.  Both scenarios require a VSCode extension, and for all cases, there's an instance of VSCode-remote server running (where VSCode is connecting).  

I think all these cases need to be handled similarly to `remote containers`

Reply:
The checks here are to force us down the device code path when we assume we are doomed to fail trying to log in using a web-browser.

We could consider trying to be more like `az` here and trying the browser login first, and fall back to device-code login if that returns an error?  What do you think?

Reply:
And unfortunately `az login` fails in certain cases like WSL with `w3m` installed (CLI interactive browser with JS disabled), in a browser-based codespace where localhost redirection is needed.

The heuristic simply doesn't work in all-cases, so if we did want to apply a heuristic, it would be after our pre-determined outcomes.

@vhvb1989 maybe you should test this change with your setup and see what works and what doesn't and maybe propose an improvement? It'd be great if we can cover most scenarios now that we own `login`